backpack/extensions/module_extension.py

```diff
@@ -12,98 +24,156 @@ class ModuleExtension:
 
-    def __init__(self, params=None):
-        """
-        Parameters
-        ----------
-        params: [str]
-            List of module parameters that need special treatment.
-            for each param `p` in the list, instances of the extended module `m`
-            need to have a field `m.p` and the class extending `ModuleExtension`
-            need to provide a method with the same signature as the `backprop`
-            method.
-            The result of this method will be saved in the savefield of `m.p`.
+    def __init__(self, params: List[str] = None):
+        """Initialization.
+
+        Args:
+            params: List of module parameters that need special treatment.
+                for each param `p` in the list, instances of the extended module `m`
+                need to have a field `m.p` and the class extending `ModuleExtension`
+                need to provide a method with the same signature as the `backprop`
+                method.
+                The result of this method will be saved in the savefield of `m.p`.
+
+        Raises:
+            NotImplementedError: if child class doesn't have a method for each parameter
         """
         if params is None:
             params = []
 
         self.__params = params
 
         for param in self.__params:
-            extFunc = getattr(self, param, None)
-            if extFunc is None:
+            if hasattr(self, param) is False:
                 raise NotImplementedError
 
-    def backpropagate(self, ext, module, g_inp, g_out, bpQuantities):
-        """
-        Main method to extend to backpropagate additional information through
-        the graph.
-
-        Parameters
-        ----------
-        ext: BackpropExtension
-            Instance of the extension currently running
-        module: torch.nn.Module
-            Instance of the extended module
-        g_inp: [Tensor]
-            Gradient of the loss w.r.t. the inputs
-        g_out: Tensor
-            Gradient of the loss w.r.t. the output
-        bpQuantities:
-            Quantities backpropagated w.r.t. the output
+    def backpropagate(
+        self,
+        extension: BackpropExtension,
+        module: Module,
+        g_inp: Tuple[Tensor],
+        g_out: Tuple[Tensor],
+        bpQuantities: Any,
+    ) -> Any:
+        """Backpropagation of additional information through the graph.
+
+        Args:
+            extension: Instance of the extension currently running
+            module: Instance of the extended module
+            g_inp: Gradient of the loss w.r.t. the inputs
+            g_out: Gradient of the loss w.r.t. the output
+            bpQuantities: Quantities backpropagated w.r.t. the output
 
         Returns
-        -------
-        bpQuantities:
             Quantities backpropagated w.r.t. the input
         """
         warnings.warn("Backpropagate has not been overwritten")
 
-    def apply(self, ext, module, g_inp, g_out):
-        """
+    def apply(
```

Comment:
Let's use python's built-in `__call__` method to make an object callable: Rename into `__call__`. Replace `module_extension.apply(...)` by `module_extension(...)`.

Reply:
Nice idea. Implemented it.